fleetspeak/src/server/internal/cache/cache.go

```diff
@@ -0,0 +1,114 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/google/fleetspeak/fleetspeak/src/common"
+	"github.com/google/fleetspeak/fleetspeak/src/server/db"
+)
+
+// Controls cache timing behavior, variable only to support unit tests.
+var (
+	maxAge         = 30 * time.Second
+	expireInterval = 5 * time.Minute
+)
+
+// Client is a cache of recently connected clients.
+type Client struct {
+	m    map[common.ClientID]*clientEntry
+	l    sync.RWMutex
+	stop chan struct{}
+}
+
+type clientEntry struct {
+	u time.Time
+	d *db.ClientData
+}
+
+// NewClient returns a new cache of client data.
+func NewClient() *Client {
+	ret := &Client{
+		m:    make(map[common.ClientID]*clientEntry),
+		stop: make(chan struct{}),
+	}
+	go ret.expireLoop()
+	return ret
+}
+
+// Get returns the cached client data, if there is sufficiently fresh data in
+// the cache, otherwise nil.
+func (c *Client) Get(id common.ClientID) *db.ClientData {
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	e := c.m[id]
+	if e == nil || db.Now().Sub(e.u) > maxAge {
+		return nil
+	}
+	return e.d
+}
+
+// Update updates or sets the cached data for a particular client. If data is
+// nil, it clears the data for the client.
+func (c *Client) Update(id common.ClientID, data *db.ClientData) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	if data == nil {
+		delete(c.m, id)
+	} else {
+		c.m[id] = &clientEntry{
+			u: db.Now(),
+			d: data,
+		}
+	}
+}
+
+// Clear empties the cache, removing all entries.
+func (c *Client) Clear() {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	c.m = make(map[common.ClientID]*clientEntry)
+}
+
+// Stop releases the resources required for background cache maintenence. The
+// Cache should not be used once Stop has been called.
+func (c *Client) Stop() {
+	close(c.stop)
+}
+
+// Size returns the current size taken up by the cache, this is a count of
+// client records, some of which may be too old to be visible.
+func (c *Client) Size() int {
+	c.l.RLock()
+	defer c.l.RUnlock()
+	return len(c.m)
+}
+
+func (c *Client) expireLoop() {
```

Comment:
Making cleanup periodical is surprising to me. While this works, another obvious way that seems cleaner to me is triggering cleanup once every 100000 (or a different constant) operations, or when len(c.m) % 10000 == 0. What do you think?

Reply:
We could have all the clients in the system cached without noticing the ram increase. (100k clients, 250 bytes/client ->25MB? Our least loaded server tasks are using 1GB) This is really just to be tidy, not to keep memory size down.

A client will take a few minutes to fall out of fastpoll, this is independent of the number of operations that occur.

Adding comments to explain the design choices.

Reply:
I didn't refer to RAM. Adding a garbage collector goroutine isn't tidy - is all I'm saying. Do we need it?

Reply:
Making the cleanup time-based, when the need for cleanup tends to happen after a certain amount of time, is in my mind tidy. We could avoid the goroutine by checking the time on every op, but these are goroutines, not pthreads. They are cheap and meant to be freely used.

Reply:
I think it's the tidiest to not have a GC at all. See the comment below.